### PR TITLE
:sparkles: Added option to dry run one or all notebooks

### PR DESCRIPTION
**Pull request by Vinsho**
- Added some operation logging as well

### CODE REVIEW

File: src/odap/feature_factory/dry_run.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    dataframes, metadata = create_dataframes_and_metadata(
+        entity_primary_key, features_paths=get_list_of_selected_features()
+    )
+
+    set_fs_compatible_metadata(metadata, config)
+
+    metadata_df = create_dataframe(metadata, get_metadata_schema())
+
+    if len(dataframes) > 1:
+        join_dataframes(dataframes, join_columns=[entity_primary_key])
```
Comment (guderkar) [inline src/odap/feature_factory/dry_run.py:42]: I thin timestamp should be also join keys

File: src/odap/feature_factory/dry_run.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+FEATURE_WIDGET = "feature"
+
+
+def get_list_of_selected_features() -> List[str]:
+    feature_name = get_widget_value(FEATURE_WIDGET)
+
+    if feature_name == ALL:
+        return get_features_paths(get_workspace_api())
+
+    return [os.path.join(get_absolute_path("features"), feature_name)]
```
Comment (guderkar) [inline src/odap/feature_factory/dry_run.py:26]: how this works? feature_name might be somethink like `f1,f2,f3` thats what widget returns right?

File: src/odap/feature_factory/feature_store.py [hunk 3]
```diff
@@ -41,4 +42,6 @@ def write_df_to_feature_store(
 
     create_feature_store_table(fs, df, table_name, table_path, primary_keys, partition_columns)
 
+    logger.info("Writing data to table: %s...", table_name)
```
Comment (guderkar) [inline src/odap/feature_factory/feature_store.py:45]: use f string

File: src/odap/feature_factory/dataframes.py [hunk 4]
```diff
@@ -35,11 +69,15 @@ def join_dataframes(dataframes: List[DataFrame], join_columns: List[str]) -> Dat
     union_df = reduce(lambda df1, df2: df1.unionByName(df2, allowMissingColumns=True), dataframes)
     columns = [col for col in union_df.columns if col not in join_columns]
 
-    return (
+    logger.info("Joining %i dataframes...", len(dataframes))
```
Comment (guderkar) [inline src/odap/feature_factory/dataframes.py:72]: use f string

File: src/odap/feature_factory/dry_run.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    if len(dataframes) > 1:
+        join_dataframes(dataframes, join_columns=[entity_primary_key])
+
+    logger.info("Success. No errors found!\nMetadata Table:")
+
+    metadata_df.display()
+
+
+def create_notebook_widget():
```
Comment (guderkar) [inline src/odap/feature_factory/dry_run.py:49]: maybe move any widget logic do widgets.py?

**Review by guderkar — CHANGES_REQUESTED** [review]

File: src/odap/feature_factory/dry_run.py (hunk 2 above)
Comment (Vinsho) [reply src/odap/feature_factory/dry_run.py:26]: Right now you can select only one feature or `<all>`. If that's what you meant.

File: src/odap/feature_factory/feature_store.py (hunk 3 above)
Comment (Vinsho) [reply src/odap/feature_factory/feature_store.py:45]: I did :D but it shouted
`Use lazy % formatting in logging functions (logging-fstring-interpolation)`

File: src/odap/feature_factory/dry_run.py (hunk 2 above)
Comment (guderkar) [reply src/odap/feature_factory/dry_run.py:26]: ok

File: src/odap/feature_factory/feature_store.py (hunk 3 above)
Comment (guderkar) [reply src/odap/feature_factory/feature_store.py:45]: So let's disable this warning in pyproject.toml

File: src/odap/feature_factory/dry_run.py (hunk 5 above)
Comment (Vinsho) [reply src/odap/feature_factory/dry_run.py:49]: Well there's only one and it's quite dry_run specific.

**Review by guderkar — APPROVED** [review]

**Pull request merged**